crates/utils/src/http_client.rs

```diff
@@ -0,0 +1,845 @@
+
+        let identity =
+            Identity::from_pkcs8_pem(&cert, &key).expect("Failed to build the identity from certificate and key");
+        let certificate = Certificate::from_pem(server_cert.as_slice()).expect("Failed to add root certificate");
+
+        let client = HttpClient::builder(&mock_server.base_url())
+            .identity(identity)
+            .add_root_certificate(certificate)
+            .build()
+            .unwrap();
+
+        // Since we can't check the certificates directly, we'll just verify the client was built
+        assert_eq!(client.base_url.as_str(), mock_server.base_url() + "/");
+    }
+
+    /// # Edge Cases
+
+    /// Tests behavior with empty or problematic paths:
+    /// - Empty path segments
+    /// - Multiple slashes
+    /// - Unicode paths
+    #[test]
+    fn test_empty_path_handling() {
+        let mock_server = httpmock::MockServer::start();
+
+        let client = HttpClient::builder(&mock_server.base_url()).build().unwrap();
+
+        // Test empty path
+        let request = client.request().path("");
+        assert_eq!(request.path, "");
+
+        // Test multiple slashes
+        let request = client.request().path("//test//path//");
+        assert_eq!(request.path, "//test//path//");
+
+        // Test Unicode path
+        let request = client.request().path("/测试/路径");
+        assert_eq!(request.path, "/测试/路径");
+    }
+
+    /// Verifies handling of special characters in:
+    /// - URLs and paths
+    /// - Query parameters
+    /// - Headers
+    /// - Form data
+    #[test]
+    fn test_special_characters() {
+        let mock_server = httpmock::MockServer::start();
+
+        let client =
+            HttpClient::builder(&mock_server.base_url()).default_query_param("special", "!@#$%^&*()").build().unwrap();
+
+        let request = client
+            .request()
+            .path("/path with spaces/and#special/chars")
+            .query_param("unicode", "测试")
+            .form_text("special form", "value with spaces");
+
+        assert!(request.form.is_some());
+    }
+
+    /// Tests client behavior with:
+    /// - Large body content
+    /// - Large file uploads
+    /// - Memory usage monitoring
+    #[tokio::test]
+    async fn test_large_payload_handling() {
+        let mock_server = httpmock::MockServer::start();
+
+        let client = HttpClient::builder(&mock_server.base_url()).build().unwrap();
+
+        // Create a large body string
+        let large_body = "x".repeat(1024 * 1024); // 1MB string
+        let request = client.request().method(Method::POST).body(&large_body);
+
+        assert!(request.body.is_some());
+        assert_eq!(request.body.unwrap().len(), (1024 * 1024) + 2);
+    }
+
+    /// Tests body serialization functionality:
+    /// - JSON serialization of different types
+    /// - Struct serialization
+    /// - Error handling for serialization failures
+    #[test]
+    fn test_request_builder_body_serialization() {
```

Comment:
let's move these edge cases above the integration ones?

Reply:
done